### ISSUE TITLE
New interfaces should conform with the Web Content Accessibility Guidelines 2.1 at level AA

### ISSUE THREAD

**Issue by richard015ar** [pressbooks/pressbooks-multi-institution#129]
As a Network Manager, Super Admin and Institutional Manager user, I want to be able to adequately access all the interfaces created by this plugin.

DoD: Apply necessary changes to ensure that all new interfaces conform with the Web Content Accessibility Guidelines 2.1 at level AA.

**Comment by greatislander** [comment]
## WCAG 2.1 issues

### [1.1.1 Non-text content](https://www.w3.org/WAI/WCAG21/Understanding/non-text-content)
- [x] **Institution List:** An emoji character is used as the value of the buy-in column in each institution’s row. With VoiceOver (macOS) it is announced as “cross mark” (for false) or “check mark” (for true) which doesn’t meaningfully communicate the value of this option. Ideally the value would be communicated as “yes” or “no”. You could use heroicons (e.g. `check-circle` and `x-circle`) and put the accessible text in a screen-reader-text span adjacent to the icon, e.g.:
	```html
	<i class="pb-heroicons-check-circle" aria-hidden-"true"></i>
    <span class="screen-reader-text"><?php __('yes', 'pressbooks-multi-institution') ?></span>
	```
- [x] **Dashboard:** The “View book list”, “View user list” and “Explore stats” links all have icons which are announced in the accessible name of the link. The `<i>` tags should have `aria-hidden=“true”` added to them (as seen in pressbooks/pressbooks#3618).

### [1.3.1 Info and relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships)
- [x] **Institution List:** The last four rows in the table should not be there, as they don’t fit within the table semantics. The value of the second column in each row is the name of the institution, but the last four rows do not follow this convention. They also are not sortable or selectable for bulk actions and make the row count ambiguous (the table will be announced to screen reader users as having five rows even if it only has one institution). It would be much better to include a non-interactive summary table below the institutions table which shows the unassigned, shared network, premium member and grand totals for users and books.
- [ ] **Add/Edit Institution:** The “Institution Managers Allowed” checkbox does not have any perceivable relationship to or impact on the “Institutional Managers” field. If it allows or disallows the use of the “Institutional Managers” field, the field should be hidden or shown as appropriate (or disabled or enabled as appropriate).
- [x] **Assign Books:** The Book Administrators in the table are communicated in a way that might be confusing; they’re announced as “username, email@example.com, unassigned” if they aren’t assigned to an institution. It would be less ambiguous to have the text read “No institution assigned” instead of “unassigned”.

### [1.4.3 Contrast (minimum)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum)
- [x] **All pages:** The contrast between the text and background of active sidebar navigation items (e.g. Institutions) is not sufficient; it’s 4.47:1 but must be 4.5:1. (Note: this is a Pressbooks issue.)

### [2.4.2 Paged Titled](https://www.w3.org/WAI/WCAG21/Understanding/page-titled)
- [x] **Add/Edit Institution:** The same form is used for adding and editing institutions. When editing, the page’s title (in the `<title>` tag) still reads “Add Institution”.

### [3.2.2 On Input](https://www.w3.org/WAI/WCAG21/Understanding/on-input)
- [ ] **Add Institution:** When an institution is added, the user is redirected to an empty “Add Institution” page. It would be more intuitive for them to be redirected to the Institutions List upon successful form submission.

### [3.3.1 Error identification](https://www.w3.org/WAI/WCAG21/Understanding/error-identification)
- [x] **Add/Edit Institution:** Form fields that have an error are not visually or programatically indicated. For example, if I try to enter an email domain in use by another institution, the error message appears at the top of the page but the field itself does not reference the error message using `aria-describedby` (you can set an `aria-describedby` attribute to multiple IDs, e.g. `aria-describedby="email-domains-description email-domains-error"`). It also does not use the `aria-invalid` attribute to indicate that it is in an error state. There is also no visual indication of the field with the error. For ideas on implementing this, see: https://www.w3.org/WAI/WCAG21/Understanding/error-identification
- [x] **Add/Edit Institution:** Although the errors are displayed in  `<div>` with the alert role, they are not announced because the error messages are already in the container when the page loads (for more information, see: https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA19). One option to consider is injecting the error messages immediately after the page has loaded which should trigger the announcement to screen reader users.

### [3.3.2 Labels or instructions](https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions)
- [x] **Add/Edit Institution:** The instructions for adding email domains say that one domain should be added per line but it’s not possible to add multiple lines of text; instead you must add an email domain with the button. The instructions should be updated to reflect this.
- [x] **Add/Edit Institution:** The “Add New” button for adding email domains should have an accessible name of “Add new email domain”; without that context it’s not clear what you are adding to a non-sighted user.
- [x] **Add/Edit Institution:** The “Institutional Managers” multiple select has a label of “Test” ([see here](https://github.com/pressbooks/pressbooks-multi-institution/blob/fe98c56867b6fd0945be3476646881c750355c57/resources/views/institutions/form.blade.php#L156)). It should match the visible label.
- [x] **Add/Edit Institution:** The “Book Limit” field’s description is not programatically associated with the input, and as such is not exposed to screen reader users.
- [x] **Add/Edit Institution:** The “Book Limit” field’s description suggested entering 0 for no limit. It’s not immediately clear to this user what leaving the field blank will do.
- [x] **Add/Edit Institution:** When adding an institution, the submit button’s label, “Save Changes”, may be ambiguous. It would be clearer to label the button “Add Institution” and then label the button “Save Changes” when editing an existing institution.

### [4.1.2 Name, role, value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value)
- [x] **Institutions List:** The checkboxes in the first column (other than the top and bottom rows) have no accessible label. The label should be “Select `<institution name>`”.
- [x] **Add/Edit Institution:** When the email domains field is empty, a PHP error can be seen in the source which leads to the field not being labelled:
	```html
	<input id="domains-1" name="domains[]" type="text" value="" class="regular-text" aria-labelledby="
	Warning: Undefined variable $name in /app/web/app/uploads/pressbooks/cache/9c850ab6a1bcda1f61afd7a10233008f2560e419.php on line 106
	-label" aria-describedby="
	Warning: Undefined variable $name in /app/web/app/uploads/pressbooks/cache/9c850ab6a1bcda1f61afd7a10233008f2560e419.php on line 107
	-description">\
	```
	
### [4.1.3 Status Messages](https://www.w3.org/WAI/WCAG21/Understanding/status-messages)
- [x] **Add Institution:** Once an institution has been successfully added, a message is displayed on the “Add Institution” page indicating that the operation has succeeded. However, it is not properly announced because it is present in the `role="status"` container when the page loads (for more information, see: https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA19). One option to consider is injecting the message immediately after the page has loaded which should trigger the announcement to screen reader users.

## Other Recommendations

- [x] It might be better to make the Book Limit input a standard text input with the affordances recommended here (not a WCAG violation as is, but the suggestion is a better user experience for a wide range of users): https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/

**Comment by fdalcin** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Q4-2024-65061c7b8bd3139dea5e5373/issues/pressbooks/pressbooks-multi-institution/129?planning-poker) with Zenhub @ho-man-chan @richard015ar

**Comment by richard015ar** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Q4-2024-65061c7b8bd3139dea5e5373/issues/pressbooks/pressbooks-multi-institution/129?planning-poker) with Zenhub @tw77

**Comment by SteelWagstaff** [comment]
> Institution List: The last four rows in the table should not be there, as they don’t fit within the table semantics. The value of the second column in each row is the name of the institution, but the last four rows do not follow this convention. They also are not sortable or selectable for bulk actions and make the row count ambiguous (the table will be announced to screen reader users as having five rows even if it only has one institution). It would be much better to include a non-interactive summary table below the institutions table which shows the unassigned, shared network, premium member and grand totals for users and books.

Agree with this suggestion!

> Add/Edit Institution: The “Institution Managers Allowed” checkbox does not have any perceivable relationship to or impact on the “Institutional Managers” field. If it allows or disallows the use of the “Institutional Managers” field, the field should be hidden or shown as appropriate (or disabled or enabled as appropriate).

@greatislander when this checkbox is selected, _network managers_ are able to assign and remove institutional managers. When it is unselected, they cannot add or remove institutional managers. It has no perceivable impact for super admins, who can do this all the time, only for network managers (network managers don't see all the options on this form).


> It might be better to make the Book Limit input a standard text input with the affordances recommended here (not a WCAG violation as is, but the suggestion is a better user experience for a wide range of users): https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/

Interesting suggestion. FWIW, the book limit field is only visible to super admins (i.e. Pressbook staff) and will never be seen by clients. Not sure how important addressing this is, but appreciate learning more about inclusive best practices.

**Comment by MichelleWeremczuk** [comment]
Separation of book totals from main institution list chart [figma link](https://www.figma.com/file/ctFGUz0nF4cJb016vVg5qA/Chart-Interfaces?type=design&node-id=849%3A2179&mode=design&t=6UI3HqsR0EQ4cUUX-1)

**Comment by fdalcin** [comment]
@greatislander for status/error messages would something like this work? How can I properly test it?

```blade
@if (!empty($_POST) && isset($result['success']) && isset($result['message']))
	<div
		id="message"
		role="status"
		class="updated notice is-dismissible {{ $result['success'] ? '' : 'error' }}"
		x-data="{
			data: @js($result),
			show: false,
		}"
		x-init="$nextTick(() => show = true)"
	>
		<template x-if="show">
			<p>
				<template x-if="data.success">
					<strong>{{ __('Success', 'pressbooks-multi-institution') }}:</strong>
				</template>
				<template x-if="! data.success">
					<strong>{{ __('Error', 'pressbooks-multi-institution') }}:</strong>
				</template>

				<template x-if="data.message">
					<span x-text="data.message"></span>
				</template>
			</p>
		</template>

		<template x-if="Object.keys(data.errors).length > 0 && show">
			<ul>
				<template x-for="field in Object.keys(data.errors)">
					<li class="padding" x-bind:id="`${field}-errors`">
						<template x-for="error in data.errors[field]">
							<span x-html="error"></span>
						</template>
					</li>
				</template>
			</ul>
		</template>
	</div>
@endif
```

**Comment by greatislander** [comment]
> @greatislander for status/error messages would something like this work? How can I properly test it?

Hi @fdalcin,

I pasted your sample into a blank HTML file like this (I simplified and removed all the PHP code):

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<meta charset="UTF-8">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		<title>Alert Test</title>
		<script src="https://unpkg.com/alpinejs" defer></script>
	</head>
	<body>
		<!-- Your code here -->
	</body>
</html>
```

Then I loaded it in Safari and turned on VoiceOver.

### Expected behaviour

I refresh the page and hear the alert message announced by VoiceOver.

### Actual behaviour

No announcement.

However, I tried a different approach— instead of using `$nextTick` I did this:

```
x-init="setTimeout(function() {show = true}, 200)"
```

And it seems to work as expected! Here's a Codepen: https://codepen.io/greatislander/full/gOyzXxQ

**Comment by tw77** [comment]
Testing on dev.pressbooks.network, the implemented changes are working as expected without issues.

I notice that this isn't checked as complete above and isn't implemented:

> Add Institution: When an institution is added, the user is redirected to an empty “Add Institution” page. It would be more intuitive for them to be redirected to the Institutions List upon successful form submission.

I'll bring this up on Monday at standup since it may be already known and intentionally not included in the changes made here. I see how this redirect to the Institutions List would be beneficial, though. Something for a future sprint?

This one has been checked as complete above:

> The “Book Limit” field’s description suggested entering 0 for no limit. It’s not immediately clear to this user what leaving the field blank will do.

But the description for the "Book Limit" field hasn't been changed. Maybe it wasn't changed here because the field is only shown to super admins and won't be shown to clients? I'll ask on Monday and re-open if needed.
